components/supervisor/frontend/src/ide/supervisor-service-client.ts

```diff
@@ -84,13 +78,45 @@ export class SupervisorServiceClient {
         if (kind == "supervisor") {
             wait = "";
         }
+
+        // setup a timeout, which is meant to re-establish the connection every 5 seconds
+        const timeout = new Timeout(5000);
+
+        // track whenever a) we are done, or b) we try to connect (again)
+        const trackCheckReady = (p: { aborted?: boolean }, err?: any): void => {
+            const props: Record<string, string> = {
+                component: "supervisor-frontend",
+                instanceId: this.serviceClient.latestInfo?.instanceId ?? "",
+                userId: this.serviceClient.latestInfo?.loggedUserId ?? "",
+                readyKind: kind,
+            };
+            if (err) {
+                props.errorName = err.name;
+                props.errorStack = err.message ?? String(err);
+            }
+
+            props.aborted = String(!!p.aborted);
+            props.wait = wait;
+
+            this.serviceClient.trackEvent({
+                event: "supervisor_check_ready",
+                properties: props,
+            });
+        };
+
+        let isError = false;
         try {
+            timeout.reset();
```

Comment:
@mustard-mh I'd love to rollout this whole change behind a feature flag, but I'm not sure what the best way is to retrieve one.

I thought about a way to retrieve values from the dashboard (to have a central place where we fetch it, and also so it's already loaded once we are in supervisor frontend). but I'm not entirely sure how to go about it.

I could see a "request -> response" pattern between export class FrontendDashboardServiceClient implements IDEFrontendDashboardService.IClient {
    public latestInfo!: IDEFrontendDashboardService.Info;
    private credentialsToken?: Uint8Array;

    private readonly onDidChangeEmitter = new Emitter<IDEFrontendDashboardService.Info>();
    readonly onInfoUpdate = this.onDidChangeEmitter.event;

    private readonly onOpenBrowserIDEEmitter = new Emitter<void>();
    readonly onOpenBrowserIDE = this.onOpenBrowserIDEEmitter.event;

    private readonly onWillRedirectEmitter = new Emitter<void>();
    readonly onWillRedirect = this.onWillRedirectEmitter.event;

    private resolveInit!: () => void;
    private initPromise = new Promise<void>((resolve) => (this.resolveInit = resolve));

    private version?: number;

    constructor(private serverWindow: Window) {
        window.addEventListener("message", (event: MessageEvent) => {
            if (event.origin !== serverUrl.url.origin) {
                return;
            }
            if (IDEFrontendDashboardService.isInfoUpdateEventData(event.data)) {
                metricsReporter.updateCommonErrorDetails({
                    userId: event.data.info.loggedUserId,
                    ownerId: event.data.info.ownerId,
                    workspaceId: event.data.info.workspaceID,
                    instanceId: event.data.info.instanceId,
                    instancePhase: event.data.info.statusPhase,
                });
                this.version = event.data.version;
                this.latestInfo = event.data.info;
                if (event.data.info.credentialsToken?.length > 0) {
                    this.credentialsToken = Uint8Array.from(atob(event.data.info.credentialsToken), (c) =>
                        c.charCodeAt(0),
                    );
                }
                this.resolveInit();
                this.onDidChangeEmitter.fire(this.latestInfo);
            }
            if (IDEFrontendDashboardService.isRelocateEventData(event.data)) {
                this.onWillRedirectEmitter.fire();
                window.location.href = event.data.url;
            }
            if (IDEFrontendDashboardService.isOpenBrowserIDE(event.data)) {
                this.onOpenBrowserIDEEmitter.fire(undefined);
            }
        });
    }
    initialize(): Promise<void> {
        return this.initPromise;
    }

    decrypt(str: string): string {
        if (!this.credentialsToken) {
            throw new Error("no credentials token available");
        }
        const obj = JSON.parse(str);
        if (!isSerializedEncryptedData(obj)) {
            throw new Error("incorrect encrypted data");
        }
        const data = {
            ...obj,
            iv: Buffer.from(obj.iv, "base64"),
            tag: Buffer.from(obj.tag, "base64"),
        };
        const decipher = crypto.createDecipheriv("aes-256-gcm", this.credentialsToken, data.iv);
        decipher.setAuthTag(data.tag);
        const decrypted = decipher.update(data.encrypted, "hex", "utf8");
        return decrypted + decipher.final("utf8");
    }

    encrypt(content: string): string {
        if (!this.credentialsToken) {
            throw new Error("no credentials token available");
        }
        const iv = crypto.randomBytes(12);
        const cipher = crypto.createCipheriv("aes-256-gcm", this.credentialsToken, iv);
        let encrypted = cipher.update(content, "utf8", "hex");
        encrypted += cipher.final("hex");
        const tag = cipher.getAuthTag();
        return JSON.stringify({
            iv: iv.toString("base64"),
            tag: tag.toString("base64"),
            encrypted,
        });
    }

    isEncryptedData(content: string): boolean {
        try {
            const obj = JSON.parse(content);
            return isSerializedEncryptedData(obj);
        } catch (e) {
            return false;
        }
    }

    trackEvent(msg: RemoteTrackMessage): void {
        const debugWorkspace = workspaceUrl.debugWorkspace;
        msg.properties = { ...msg.properties, debugWorkspace };
        this.serverWindow.postMessage(
            { type: "ide-track-event", msg } as IDEFrontendDashboardService.TrackEventData,
            serverUrl.url.origin,
        );
    }

    activeHeartbeat(): void {
        this.serverWindow.postMessage(
            { type: "ide-heartbeat" } as IDEFrontendDashboardService.HeartbeatEventData,
            serverUrl.url.origin,
        );
    }

    setState(state: IDEFrontendDashboardService.SetStateData): void {
        this.serverWindow.postMessage(
            { type: "ide-set-state", state } as IDEFrontendDashboardService.SetStateData,
            serverUrl.url.origin,
        );
    }

    // always perfrom redirect to dekstop IDE on gitpod origin
    // to avoid confirmation popup on each workspace origin
    openDesktopIDE(url: string): void {
        this.serverWindow.postMessage(
            { type: "ide-open-desktop", url } as IDEFrontendDashboardService.OpenDesktopIDE,
            serverUrl.url.origin,
        );
    }
} and [here](https://github.com/gitpod-io/gitpod/blob/819244447766f4d6e4f2dcb1db239a5a224756ba/components/dashboard/src/service/service.tsx#L187-L203).

WDYT? Is there anything easier I'm missing? :thinking:

Reply:
I did not want to spread how we create the connection, and added some request/response messages on the window protocol we have going here: https://github.com/gitpod-io/gitpod/pull/20175/commits/9faa68a15fff3f33c4dddfc187994e7e8664aa03